fix: vLLM engine graceful shutdown to prevent VRAM leaks on crash

WaelBKZ opened the pull request:
#### Overview:

Fixes VRAM memory leak when vLLM engine crashes. Added proper engine shutdown to free GPU memory.

#### Details:

- Added `_shutdown_engine()` method to `VllmEngineMonitor` 
- Calls `engine_client.shutdown()` before runtime shutdown on `EngineDeadError`
- Handles shutdown failures gracefully

#### Where should the reviewer start?

Check `components/backends/vllm/src/dynamo/vllm/engine_monitor.py` - the `_shutdown_engine()` method and its usage in the exception handler.

#### Next steps:

Other components have similar `EngineDeadError` exceptions. I can add support for these as well.

#### Related Issues:

- Closes GitHub issue: https://github.com/ai-dynamo/dynamo/issues/2894

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **New Features**
  * Adds continuous background health monitoring for the model-serving engine with periodic checks to detect failures quickly.
  * Automatically performs safe cleanup and shuts down the runtime on fatal engine failures, preventing hangs and resource leaks.
  * Improves resilience and reduces manual intervention; unresponsive sessions terminate promptly, enabling quicker restarts.
  * Enhances failure visibility with clearer logs during engine issues, aiding faster troubleshooting and recovery.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

WaelBKZ committed:
fix(engine_monitor): add proper vLLM engine cleanup to prevent VRAM leaks on crash

- Add _shutdown_engine() method to VllmEngineMonitor class
- Call engine shutdown before runtime shutdown on EngineDeadError
- Prevents GPU memory leaks when vLLM engine crashes

Fixes #2894

Signed-off-by: Wael Boukhobza <wawa_wael@live.fr>

WaelBKZ committed:
Merge branch 'main' into fix/vllm-engine-monitor-cleanup-2894

grahamking commented on components/backends/vllm/src/dynamo/vllm/engine_monitor.py:
```diff
@@ -43,13 +43,26 @@ def __init__(self, runtime: DistributedRuntime, engine_client: AsyncLLM):
+        try:
+            logger.info("Shutting down vLLM engine to free GPU memory...")
+            self.engine_client.shutdown()
+            logger.info("vLLM engine shutdown completed")
```
Can you remove the log statements? That will prevent on every single normal shutdown, and I feel like we already have too many "shutting down xyz" logs.

grahamking commented on components/backends/vllm/src/dynamo/vllm/engine_monitor.py:
```diff
@@ -43,13 +43,26 @@ def __init__(self, runtime: DistributedRuntime, engine_client: AsyncLLM):
                 await asyncio.sleep(HEALTH_CHECK_INTERVAL)
             except EngineDeadError as e:
                 logger.error(f"vLLM AsyncLLM health check failed: {e}")
+                logger.warning("Initiating engine cleanup.")
```
This is a good example of too many logs statements! :-)  Remove the new warning.

WaelBKZ committed:
Add timeout protection to vLLM engine shutdown

Use SIGALRM to prevent hanging during engine shutdown on crash scenarios.
Cleaned up log statements and removed unused imports.

Signed-off-by: Wael Boukhobza <wawa_wael@live.fr>

WaelBKZ commented on components/backends/vllm/src/dynamo/vllm/engine_monitor.py:
```diff
@@ -0,0 +1,70 @@
+            logger.info("vLLM engine shutdown completed")
+        except Exception as e:
+            logger.warning(f"Engine shutdown method failed: {e}")
+
```
Thanks Graham! So unlike what @coderabbitai is suggesting, `shutdown` is synchronous and needs to run on the main thread ([AsyncLLM shutdown docs](https://docs.vllm.ai/en/latest/api/vllm/v1/engine/async_llm.html?h=shutdown#vllm.v1.engine.async_llm.AsyncLLM.shutdown)).

So to circumvent that, I implemented the timeout behavior through signals in the latest commit https://github.com/ai-dynamo/dynamo/pull/2898/commits/729ac9769221057342231323b8dcb4072829a668; I tested it again, and I also tested the negative case where we timeout (with a very low timeout duration). I also cleaned up the chatty logs

WaelBKZ committed:
Merge branch 'main' into fix/vllm-engine-monitor-cleanup-2894

WaelBKZ committed:
Fix code formatting

Signed-off-by: Wael Boukhobza <wawa_wael@live.fr>

WaelBKZ committed:
Merge branch 'main' into fix/vllm-engine-monitor-cleanup-2894

WaelBKZ committed:
Merge remote-tracking branch 'upstream/main' into fix/vllm-engine-monitor-cleanup-2894

WaelBKZ committed:
Merge branch 'main' into fix/vllm-engine-monitor-cleanup-2894

grahamking commented:
/ok to test f24fead
